service/src/DCGMDevicePool.cpp

```diff
@@ -0,0 +1,186 @@
+
+        char geopm_group[] = "geopm_field_group";
+        result = dcgmFieldGroupCreate(m_dcgm_handle, sizeof(m_dcgm_field_ids)/sizeof(m_dcgm_field_ids[0]), &m_dcgm_field_ids[0],
+                                      geopm_group, &m_field_group_id);
+
+        //Retry case
+        if (result == DCGM_ST_DUPLICATE_KEY) {
+#ifdef GEOPM_DEBUG
+            std::cerr << "DCGMDevicePool::" << std::string(__func__)  <<
+                         ": Duplicate field group found. " << std::endl;
+#endif
+            result = dcgmFieldGroupDestroy(m_dcgm_handle, m_field_group_id);
+            check_dcgm_result(result, "Error destroying DCGM geopm_fields group.", __LINE__);
+            result = dcgmFieldGroupCreate(m_dcgm_handle, sizeof(m_dcgm_field_ids)/sizeof(m_dcgm_field_ids[0]), &m_dcgm_field_ids[0],
+                                          geopm_group, &m_field_group_id);
+            check_dcgm_result(result, "Error re-creating DCGM geopm_fields group.", __LINE__);
+        }
+        else {
+            check_dcgm_result(result, "Error creating DCGM geopm_fields group.", __LINE__);
+        }
+
+        // Note: Currently we are using dcgmWatchFields, but may transition to using
+        //       dcgmProfWatchFields at a later date
+        result = dcgmWatchFields(m_dcgm_handle, DCGM_GROUP_ALL_GPUS, m_field_group_id, m_update_freq,
+                                 m_max_keep_age, m_max_keep_sample);
+
+        check_dcgm_result(result, "Error setting default watch field configuration.", __LINE__);
+
+        for (auto dcgm_field_id : m_dcgm_field_ids) {
+            DcgmFieldGetById(dcgm_field_id);
+            if (DcgmFieldGetById(dcgm_field_id)->fieldType != DCGM_FT_DOUBLE) {
+                throw Exception("DCGMDevicePool::" + std::string(__func__) + ": "
+                                "DCGM Field ID " + std::to_string(dcgm_field_id) +
+                                " field type is not double",
+                                GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+            }
+        }
+    }
+
+    DCGMDevicePoolImp::~DCGMDevicePoolImp()
+    {
+        //Shutdown DCGM
+        dcgmStatusDestroy(NULL);
+        dcgmFieldGroupDestroy(m_dcgm_handle, m_field_group_id);
+        dcgmGroupDestroy(m_dcgm_handle, DCGM_GROUP_ALL_GPUS);
+    }
+
+    void DCGMDevicePoolImp::check_dcgm_result(const dcgmReturn_t result, const std::string error, const int line)
+    {
+        if (result != DCGM_ST_OK) {
+            throw Exception("DCGMDevicePool::" + std::string(__func__) + ": "
+                            + error + ": " + errorString(result),
+                            GEOPM_ERROR_INVALID, __FILE__, line);
+        }
+    }
+
+    int DCGMDevicePoolImp::dcgm_device() const {
+        return m_dcgm_dev_count;
+    }
+
+    double DCGMDevicePoolImp::field_value(int accel_idx, int geopm_field_id) const {
+        double result;
+        result = m_dcgm_field_values.at(accel_idx).at(geopm_field_id).value.dbl;
```

Comment:
Can also return NAN if .status is not DCGM_ST_OK. Hopefully doesn't happen now that we have the lower update rate, but might be good in case anyone changes the rate.